### 3 задание

In [1]:
import pandas as pd
import csv
from itertools import combinations
from collections import Counter as count

In [2]:
df = pd.read_csv('transactions.csv', delimiter = ';')        #представляем csv-файл в виде таблицы
df

PROD_CODE        BASKET_ID
0      PRD0900725  994106500000895
1      PRD0901527  994106500000895
2      PRD0901977  994106500000895
3      PRD0902163  994106500000895
4      PRD0902394  994106500000895
...           ...              ...
27802  PRD0903678  994106500835025
27803  PRD0900902  994106500835025
27804  PRD0900202  994106500835026
27805  PRD0900676  994106500835026
27806  PRD0903678  994106500835026

[27807 rows x 2 columns]

In [3]:
product_basket = {}                                           #создаем словарь, где ключ-номер корзинки
for i in range(len(df)):
    product_basket.update({df['BASKET_ID'][i]: []})

In [4]:
for i in range(len(df)):                                       #добавляем в корзинки продукты
    product_basket[df['BASKET_ID'][i]].append(df['PROD_CODE'][i])

In [5]:
len(product_basket)  #кол-во корзинок

4336

In [6]:
doubletons = []                         #создаем даблтоны
for key in product_basket:
    temp = combinations(product_basket[key], 2)
    for i in list(temp):
        doubletons.append(i)

In [7]:
len(doubletons)  #кол-во даблтонов

170850

In [8]:
prod_list = df['PROD_CODE'].values   #создаем лист продуктов

In [9]:
product_amount = dict(count(prod_list))  #создаем словарь уникальных продуктов и их кол-ва в общей таблице

In [10]:
len(product_amount)  #кол-во уникальных продуктов с кол-вом

3872

In [11]:
uniq_prod = list(product_amount.keys())  #уникальные продукты, создаю второй, потому что из прошлого будем удалять

In [12]:
len(uniq_prod)

3872

In [13]:
unfrequent_prod = []

In [14]:
for i in range(len(product_amount)):
    if product_amount[uniq_prod[i]] < 3:
        product_amount.pop(uniq_prod[i])
        unfrequent_prod.append(uniq_prod[i])      #удаляем не частотные продукты из списка с их кол-вом,
                                                  #добавляем в список не частотных

In [15]:
len(unfrequent_prod)   #кол-во нечастотных продуктов

1295

In [16]:
len(product_amount)    #кол-во частотных

2577

In [17]:
len(uniq_prod)       #кол-во уникальных

3872

In [18]:
#обозначим для продуктов id
uniq_prod_with_id = {}                                          
for i in range(len(uniq_prod)):
    uniq_prod_with_id.update({uniq_prod[i]: i+1})

In [19]:
def hashing(tuple):         #метод для хэширования, если во всех бинарных картах соблюдается уровень поддержки,                 
    key = None;             #то возвращаем 1(частотный даблтон), если нет-0
    hash1 = (uniq_prod_with_id[tuple[0]] + uniq_prod_with_id[tuple[1]]) % len(uniq_prod_with_id)
    hash2 = (uniq_prod_with_id[tuple[0]] + uniq_prod_with_id[tuple[1]] * 2) % len(uniq_prod_with_id)
    hash3 = (uniq_prod_with_id[tuple[0]] + uniq_prod_with_id[tuple[1]] * 3) % len(uniq_prod_with_id)
    if (hash1 > 3) and (hash2 > 3) and (hash3 > 3):
        key = 1
    else:
        key = 0
    return key

In [20]:
binary_map = {}           #cоздаем словарь с даблтонами и их значениями
for i in range(len(doubletons)):
    binary_map[doubletons[i]] = hashing(doubletons[i])

In [21]:
len(binary_map)   #количество уникальных даблтонов

143695

In [22]:
key_list = list(binary_map.keys())               #убираем нечастотные наборы
for i in range(len(binary_map)):
    if (binary_map[key_list[i]] == 0):
        binary_map.pop(key_list[i])

In [23]:
len(binary_map)

143347

In [24]:
new_key_list = list(binary_map.keys())      #убираем наборы, в которых содержатся нечастотные элементы из листа unfrequent_prod
for i in range(len(new_key_list)):
    for j in range(len(unfrequent_prod)):
        if new_key_list[i][0] == unfrequent_prod[j] or new_key_list[i][1] == unfrequent_prod[j]:
            try:
                binary_map.pop(tuple(new_key_list[i]))
            except:
                pass    

In [25]:
len(binary_map)      #кол-во частотных наборов:)

119906

##### Начальные данные: продукты = 27807, корзинки = 27807
##### Кол-во уникальных корзинок = 4336
##### Кол-во даблтонов = 170850
##### Кол-во уникальных продуктов = 3872 Из них частотны = 2577 Не частотны = 1295
##### Кол-во уникальных даблтонов = 143695
##### Кол-во частотных даблтонов = 119906